R/tree.R

```diff
@@ -0,0 +1,406 @@
+#'
+#' @examples
+#'
+#' x <- list(
+#'   list(id = "a", val = 2),
+#'   list(
+#'     id = "b",
+#'     val = 1,
+#'     children = list(
+#'       list(id = "b1",val = 2.5),
+#'       list(
+#'         id = "b2",
+#'         val = 8,
+#'         children = list(
+#'           list(id = "b21", val = 4)
+#'         )
+#'       )
+#'     )
+#'   ),
+#'   list(
+#'     id = "c",
+#'     val = 8,
+#'     children = list(
+#'       list(id = "c1"),
+#'       list(id = "c2", val = 1)
+#'     )
+#'   )
+#' )
+#'
+#' # Basic usage
+#' tree(x)
+#'
+#' # Even cleaner output can be achieved by not printing indices
+#' tree(x, index_unnamed = FALSE)
+#'
+#' # Limit depth if object is potentially very large
+#' tree(x, max_depth = 2)
+#'
+#' # You can customize how the values and classes are printed if desired
+#' tree(x, val_printer = function(x){paste0("_", x, "_")})
+#'
+#' @export
+tree <- function(
+  x,
+  ...,
+  index_unnamed = FALSE,
+  max_depth = 10L,
+  max_length = 1000L,
+  show_environments = TRUE,
+  hide_scalar_types = TRUE,
+  val_printer = crayon::blue,
+  class_printer = crayon::silver,
+  show_attributes = FALSE,
+  remove_newlines = TRUE,
+  tree_chars = box_chars()
+){
+  ellipsis::check_dots_empty()
+
+  # Pack up the unchanging arguments into a list and send to tree_internal
+  termination_type <- tree_internal(
+    x,
+    opts = list(
+      index_unnamed = index_unnamed,
+      max_depth = max_depth,
+      max_length = max_length,
+      show_envs = show_environments,
+      hide_scalar_types = hide_scalar_types,
+      val_printer = val_printer,
+      class_printer = class_printer,
+      show_attributes = show_attributes,
+      remove_newlines = remove_newlines,
+      tree_chars = tree_chars
+    )
+  )
+  if (termination_type == "early") {
+    cat("...", "\n")
+  }
+
+  invisible(x)
+}
+
+# Tree printing internal function
+#
+# This is the internal function for the main tree printing code. It wraps the
+# static options arguments from the user-facing `tree()` into a single opts
+# list to make recursive calls cleaner. It also has arguments that as it is
+# called successively but the end-user shouldn't see or use.
+tree_internal <- function(
+  x,
+  x_id = NULL,
+  branch_hist = character(0),
+  opts,
+  attr_mode = FALSE,
+  counter_env = rlang::new_environment(
+    data = list(n_printed = 0, envs_seen = c())
+  )
+) {
+  counter_env$n_printed <- counter_env$n_printed + 1
+  # Stop if we've reached the max number of times printed desired
+  if (counter_env$n_printed > opts$max_length) {
+    return("early")
+  }
+
+  # Since self-loops can occur in environments check to see if we've seen any
+  # environments before
+  already_seen <- rlang::is_environment(x) &&
+    any(vapply(counter_env$envs_seen, identical, x, FUN.VALUE = NA))
+
+  if (!already_seen && rlang::is_environment(x)) {
+    # If this environment is new, add it to the seen
+    counter_env$envs_seen[[length(counter_env$envs_seen) + 1]] <- x
+  }
```

Comment:
I think it would be clearer to have:

```R
  already_seen <- FALSE
  if (is_environment(x)) {
	already_seen <- any(vapply(counter_env$envs_seen, identical, x, FUN.VALUE = NA))
    if (!already_seen) {
      counter_env$envs_seen[[length(counter_env$envs_seen) + 1]] <- x
    }
  }
```

Reply:
I've recently been using `NA`/`NA_character_`/`NA_integer_` etc., because the literal value has less overhead than a function call. The drawback is that for logicals, it's `NA` instead of `NA_logical_`, so it's not as obvious what the type is.

Reply:
7c5d46e